viaMover/web-app
# [Feature] Debit card

**Context**
* Debit card module is to be introduced

**What was done**
* Added .editorconfig for the whole project
* Added EURS-WETH price helper
* Added pages (manage -- active, pending, set-email, etc)
* Added Vuex store
* Added vue-the-mask package
* Added form step guard (to be used later for refactoring)
* Added localized texts
* Updated nav items to be given `active` class instead of `exact-active` on condition
* Added new assets
* Added `SearchSkinModal` (a modal to be used the same way as `SearchTokenModal` but for skins)
* Enabled CSS source maps
* Added two new feature flags (to be able to perform rolling updates): `isDebitCardTopUpEnabled`, `isDebitCardChangeSkinEnabled`. Both are `false` atm and they should be

**How to test**
* Enable `isDebitCardEnabled` feature flag
* There should be a specific banner on the home page for the card (if only feature flag is `true`, the module is initialized and card state is `order_now`)
* Card navigation item in the home page nav menu should be clickable
* The rest of the flow is described in `TODO / Notes` section
* Optionally, one could test `Change Skin` logic and the boilerplate of `Top Up`

**TODO / Notes**
* The card module state machine manages its state by these rules:
  * If there is a hash stored, use it to identify the user and their debit card account
    * If a hash is invalid (or at least API tells us so), drop the persist item and repeat the whole process from the beginning (should not cause logical loop here)
    * Else map info response to the appropriate Vuex state to control displayable entities and flow
    
      If the state is `order_now` then proceed with the order form
      If the state is `validate_phone` then proceed with the phone validation form. The optional step is `change_phone` which triggers the service call to append a new phone number to the Trastra account. Succeeding this step will return the user to the order flow, `validate_phone` step
      Otherwise, consider the card to be active, display all the info we have at the moment. Nothing fancy here, just some card data (until the next debit card update is released)
  * Else request a hash to be created (display a form with email field)
    * Make a signature request to bind an address with an email, create persisted entry, use email for ordering if needed, otherwise never ask for email for about 1 month of the inactive period
    * Send this signature to the service API to identify the user so only they will be capable of fetching their data, nobody else won't
  * If the state can be recovered from local info, do it. Otherwise, go to the API
  * Every API request returns a valid status which could be used to modify module internal state
